x-pack/plugins/files/server/file_client/file_client.ts

```diff
@@ -0,0 +1,191 @@
+}
+
+export type UploadOptions = Omit<BlobUploadOptions, 'id'>;
+
+/**
+ * Wraps the {@link FileMetadataClient} and {@link BlobStorageClient} client
+ * to provide basic file CRUD functionality.
+ *
+ * For now this is just a shallow type of the implementation for export purposes.
+ */
+export interface FileClient {
+  /** See {@link FileMetadata.FileKind}.  */
+  fileKind: string;
+
+  /**
+   * See {@link FileMetadataClient.create}.
+   *
+   * @param arg - Arg to create a file.
+   * */
+  create(arg: CreateArgs): ReturnType<FileMetadataClient['create']>;
+
+  /**
+   * See {@link FileMetadataClient.get}
+   *
+   * @param arg - Argument to get a file
+   */
+  get: FileMetadataClient['get'];
+
+  /**
+   * {@link FileMetadataClient.update}
+   *
+   * @param arg - Argument to get a file
+   */
+  update: FileMetadataClient['update'];
+
+  /**
+   * Delete a file.
+   * @param arg - Argument to delete a file
+   */
+  delete(arg: DeleteArgs): Promise<void>;
+
+  /**
+   * See {@link BlobStorageClient.delete}
+   *
+   * @param id - Argument to delete a file
+   */
+  deleteContent: BlobStorageClient['delete'];
+
+  /**
+   * See {@link FileMetadataClient.list}
+   *
+   * @param arg - Argument to list files
+   */
+  list: FileMetadataClient['list'];
+
+  /**
+   * See {@link FileMetadataClient.find}.
+   *
+   * @param arg - Argument to find files
+   */
+  find: FileMetadataClient['find'];
+
+  /**
+   * See {@link BlobStorageClient.upload}
+   *
+   * @param id - Readable stream to upload
+   * @param rs - Readable stream to upload
+   * @param opts - Argument for uploads
+   */
+  upload(id: string, rs: Readable, opts?: UploadOptions): ReturnType<BlobStorageClient['upload']>;
+
+  /**
+   * See {@link BlobStorageClient.download}
+   *
+   * @param args - to download a file
+   */
+  download: BlobStorageClient['download'];
+}
+export class FileClientImpl implements FileClient {
+  constructor(
+    private fileKindDescriptor: FileKind,
+    private readonly metadataClient: FileMetadataClient,
+    private readonly blobStorageClient: BlobStorageClient
+  ) {}
+
+  public get fileKind(): string {
+    return this.fileKindDescriptor.id;
+  }
+
+  public create = async ({
+    id,
+    metadata,
+  }: CreateArgs): ReturnType<FileMetadataClient['create']> => {
+    return this.metadataClient.create({
+      id: id || cuid(),
+      metadata: {
+        FileKind: this.fileKind,
+        ...metadata,
+      },
+    });
+  };
+
+  public get: FileMetadataClient['get'] = async (arg) => {
+    return this.metadataClient.get(arg);
+  };
+
+  public update: FileMetadataClient['update'] = (arg) => {
+    return this.metadataClient.update(arg);
+  };
+
+  public find: FileMetadataClient['find'] = (arg) => {
+    return this.metadataClient.find(arg);
+  };
+
+  public async delete({ id, hasContent = true }: DeleteArgs) {
+    if (hasContent) await this.blobStorageClient.delete(id);
+    return this.metadataClient.delete({ id });
```

Comment:
Before deleting the blob contents we should mark file metadata as deleted.

```suggestion
  public async delete({ id, hasContent = true }: DeleteArgs) {
    await this.metadataClient.markDeleted(); // file.Status = DELETED
    if (hasContent) await this.blobStorageClient.delete(id);
    return this.metadataClient.delete({ id });
```